### PR TITLE
feature: Adds Lineage queries in artifact, context and trial components

### PR DESCRIPTION
**Pull request by yzhu0**
*Issue #, if available:*

*Description of changes:*
 feature: Add support for SageMaker lineage queries in artifact, context and trial component
- Get downstream trials in Artifact
- Get upstream trials in Artifact
- Given an S3 URI instantiate an artifact
- Get downstream datasets in DataSetArtifact
- Get upstream datasets in DataSetArtifact
- Get datasets in ImageArtifact
- Get Actions in Context
- Get ProcessingJobs in EndpointContext
- Get TransformJobs in EndpointContext
- Get TrialComponents in EndpointContext
- Get PipelineExecution in ModelPackageGroup
- Get PipelineExecution in LineageTrialCompoment
- Get upstream dataset in LineageTrialCompoment
- Get downstream models in LineageTrialCompoment

*Testing done:*
- unit test
- integration test

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backword compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by staubhp** [comment]
Please rebase so that only your commit is in the diff. This is hard to review with 15 unrelated commits

**Comment by yzhu0** [comment]
> Please rebase so that only your commit is in the diff. This is hard to review with 15 unrelated commits

rebased

File: src/sagemaker/lineage/artifact.py [hunk 1]
```diff
@@ -522,3 +584,75 @@ def endpoint_contexts(
         for vertex in query_result.vertices:
             endpoint_contexts.append(vertex.to_lineage_object())
         return endpoint_contexts
+
+    def get_upstream_datasets(
```
Comment (staubhp) [inline src/sagemaker/lineage/artifact.py:588]: Let's remove `get_` from all of these for consistency

File: src/sagemaker/lineage/artifact.py [hunk 2]
```diff
@@ -522,3 +584,75 @@ def endpoint_contexts(
...
+    """
+
+    def get_dataset(self, direction: LineageQueryDirectionEnum) -> List[Artifact]:
+        """Get artifacts representing dataset from the image artifact's lineage.
+
+        Args:
+            direction (LineageQueryDirectionEnum): The query direction.
+
+        Returns:
+            list of Artifacts: Artifacts representing an dataset.
```
Comment (staubhp) [inline src/sagemaker/lineage/artifact.py:647]: Nit: `representing a dataset`

File: src/sagemaker/lineage/lineage_trial_component.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+
+LOGGER = logging.getLogger("sagemaker")
+
+
+class LineageTrialComponent(_base_types.Record):
+    """An Amazon SageMaker, lineage trial component, which is part of a SageMaker lineage.
+
+    A trial component is a stage in a trial.
+    Trial components are created automatically within the SageMaker runtime and may not be
```
Comment (staubhp) [inline src/sagemaker/lineage/lineage_trial_component.py:38]: Trial components can be created directly.

File: src/sagemaker/lineage/artifact.py [hunk 4]
```diff
@@ -159,6 +159,57 @@ def downstream_trials(self, sagemaker_session=None) -> list:
         )
         trial_component_arns: list = list(map(lambda x: x.destination_arn, outgoing_associations))
 
+        return self._get_trial_from_trial_component(trial_component_arns)
+
+    def downstream_trials_v2(
+        self, direction: LineageQueryDirectionEnum = LineageQueryDirectionEnum.DESCENDANTS
+    ) -> list:
+        """Retrieve all downstream trial runs which that use this artifact by using lineage query.
```
[the remark this answers is not in the document]
Comment (yzhu0) [reply src/sagemaker/lineage/artifact.py]: > Nit: "Using a lineage query, retrieve all downstream trials that use this artifact"
> 
> Also change the v1 description to "Using the lineage API, retrieve all downstream trials that use this artifact"
> 
> Let's apply this wording to all of the new method descriptions.

updated all

File: src/sagemaker/lineage/artifact.py [hunk 5]
```diff
@@ -159,6 +159,57 @@ def downstream_trials(self, sagemaker_session=None) -> list:
...
+            start_arns=[self.artifact_arn],
+            query_filter=query_filter,
+            direction=direction,
+            include_edges=False,
+        )
+        trial_component_arns: list = list(map(lambda x: x.arn, query_result.vertices))
+
+        return self._get_trial_from_trial_component(trial_component_arns)
+
+    def get_upstream_trials(
```
[the remark this answers is not in the document]
Comment (yzhu0) [reply src/sagemaker/lineage/artifact.py]: > Nit: Let's keep naming consistent, ie `upstream_trials()`

updated

File: src/sagemaker/lineage/artifact.py (hunk 1 above)
Comment (yzhu0) [reply src/sagemaker/lineage/artifact.py]: > Let's remove `get_` from all of these for consistency

removed

File: src/sagemaker/lineage/artifact.py (hunk 2 above)
Comment (yzhu0) [reply src/sagemaker/lineage/artifact.py]: > Nit: `representing a dataset`

updated

File: src/sagemaker/lineage/lineage_trial_component.py (hunk 3 above)
Comment (yzhu0) [reply src/sagemaker/lineage/lineage_trial_component.py]: > Trial components can be created directly.

updated

File: src/sagemaker/lineage/artifact.py [hunk 6]
```diff
@@ -159,6 +159,57 @@ def downstream_trials(self, sagemaker_session=None) -> list:
         )
         trial_component_arns: list = list(map(lambda x: x.destination_arn, outgoing_associations))
 
+        return self._get_trial_from_trial_component(trial_component_arns)
+
+    def downstream_trials_v2(
+        self, direction: LineageQueryDirectionEnum = LineageQueryDirectionEnum.DESCENDANTS
```
Comment (staubhp) [inline src/sagemaker/lineage/artifact.py:165]: Same comment as the other PR. For methods that explicitly callout "upstream" or "downstream", don't allow direction to be set.

Can we have a `_trials()` method that has all of the shared code between `downstream_trials_v2` and `upstream_trials`?

File: src/sagemaker/lineage/artifact.py [hunk 7]
```diff
@@ -335,6 +386,17 @@ def list(
             sagemaker_session=sagemaker_session,
         )
 
+    def get_artifacts(self, s3_uri: str):
```
Comment (staubhp) [inline src/sagemaker/lineage/artifact.py:389]: `get_artifacts` -> `s3_uri_artifacts` ? 

Python doesn't support function overloading so we have to be careful about naming these too generically.

File: src/sagemaker/lineage/artifact.py [hunk 8]
```diff
@@ -522,3 +584,75 @@ def endpoint_contexts(
         for vertex in query_result.vertices:
             endpoint_contexts.append(vertex.to_lineage_object())
         return endpoint_contexts
+
+    def upstream_datasets(
+        self, direction: LineageQueryDirectionEnum = LineageQueryDirectionEnum.ASCENDANTS
```
Comment (staubhp) [inline src/sagemaker/lineage/artifact.py:589]: Same comment as above re: removing direction and moving common code to internal function

File: src/sagemaker/lineage/artifact.py [hunk 9]
```diff
@@ -522,3 +584,75 @@ def endpoint_contexts(
...
+
+
+class ImageArtifact(Artifact):
+    """A SageMaker lineage artifact representing an image.
+
+    Common model specific lineage traversals to discover how the image is connected
+    to other entities.
+    """
+
+    def dataset(self, direction: LineageQueryDirectionEnum) -> List[Artifact]:
```
Comment (staubhp) [inline src/sagemaker/lineage/artifact.py:640]: `dataset` -> `datasets`

File: src/sagemaker/lineage/context.py [hunk 10]
```diff
@@ -382,11 +405,78 @@ def training_job_arns(
             training_job_arns.append(trial_component["Source"]["SourceArn"])
         return training_job_arns
 
+    def processing_jobs(
+        self, direction: LineageQueryDirectionEnum = LineageQueryDirectionEnum.ASCENDANTS
+    ) -> List[str]:
+        """Use the lineage query to retrieve upstream processing job that use this endpoint.
```
Comment (staubhp) [inline src/sagemaker/lineage/context.py:411]: Let's remove "upstream" here given the user can change direction

File: src/sagemaker/lineage/context.py [hunk 11]
```diff
@@ -382,11 +405,78 @@ def training_job_arns(
...
+            query_filter=query_filter,
+            direction=direction,
+            include_edges=False,
+        )
+        return [vertex.to_lineage_object() for vertex in query_result.vertices]
+
+    def transform_jobs(
+        self, direction: LineageQueryDirectionEnum = LineageQueryDirectionEnum.ASCENDANTS
+    ) -> List[LineageTrialComponent]:
+        """Use the lineage query to retrieve upstream transform job that use this endpoint.
```
Comment (staubhp) [inline src/sagemaker/lineage/context.py:433]: Remove "upstream", make "transform job" plural

File: src/sagemaker/lineage/context.py [hunk 12]
```diff
@@ -382,11 +405,78 @@ def training_job_arns(
...
+
+        Args:
+            direction (LineageQueryDirectionEnum, optional): The query direction.
+
+        Returns:
+            list of str: Transform jobs.
```
Comment (staubhp) [inline src/sagemaker/lineage/context.py:439]: This is the wrong return type

File: src/sagemaker/lineage/context.py [hunk 13]
```diff
@@ -382,11 +405,78 @@ def training_job_arns(
...
+    def trial_components(
+        self, direction: LineageQueryDirectionEnum = LineageQueryDirectionEnum.ASCENDANTS
+    ) -> List[LineageTrialComponent]:
+        """Use the lineage query to retrieve upstream trial components that use this endpoint.
+
+        Args:
+            direction (LineageQueryDirectionEnum, optional): The query direction.
+
+        Returns:
+            list of str: Trial components ARNs.
```
Comment (staubhp) [inline src/sagemaker/lineage/context.py:461]: Wrong return type

File: src/sagemaker/lineage/context.py [hunk 14]
```diff
@@ -382,11 +405,78 @@ def training_job_arns(
...
+            query_filter=query_filter,
+            direction=direction,
+            include_edges=False,
+        )
+        return [vertex.to_lineage_object() for vertex in query_result.vertices]
+
+    def trial_components(
+        self, direction: LineageQueryDirectionEnum = LineageQueryDirectionEnum.ASCENDANTS
+    ) -> List[LineageTrialComponent]:
+        """Use the lineage query to retrieve upstream trial components that use this endpoint.
```
Comment (staubhp) [inline src/sagemaker/lineage/context.py:455]: Remove "upstream"

File: src/sagemaker/lineage/lineage_trial_component.py [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+        )["Tags"]
+        for tag in tags:
+            if tag["Key"] == "sagemaker:pipeline-execution-arn":
+                return tag["Value"]
+        return None
+
+    def dataset_artifacts(
+        self, direction: LineageQueryDirectionEnum = LineageQueryDirectionEnum.ASCENDANTS
+    ) -> List[Artifact]:
+        """Use the lineage query to retrieve upstream datasets that use this trial component.
```
Comment (staubhp) [inline src/sagemaker/lineage/lineage_trial_component.py:144]: Remove "upstream"

File: src/sagemaker/lineage/lineage_trial_component.py [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+            direction=direction,
+            include_edges=False,
+        )
+
+        return [vertex.to_lineage_object() for vertex in query_result.vertices]
+
+    def models(
+        self, direction: LineageQueryDirectionEnum = LineageQueryDirectionEnum.DESCENDANTS
+    ) -> List[Artifact]:
+        """Use the lineage query to retrieve downstream models that use this trial component.
```
Comment (staubhp) [inline src/sagemaker/lineage/lineage_trial_component.py:167]: Remove "downstream"

File: tests/integ/sagemaker/lineage/test_artifact.py [hunk 17]
```diff
@@ -120,6 +127,33 @@ def validate():
...
+
+def test_upstream_trials(upstream_trial_associated_artifact, trial_obj, sagemaker_session):
+    trials = upstream_trial_associated_artifact.upstream_trials()
+    assert len(trials) == 1
+    assert trial_obj.trial_name in trials
+
+
+def test_upstream_trials_v2(trial_associated_artifact, trial_obj):
+    def validate():
+        for i in range(3):
```
Comment (staubhp) [inline tests/integ/sagemaker/lineage/test_artifact.py:144]: Do we need this manual retry if we're already wrapping this function in `retry()` ?

Comment (yzhu0) [reply tests/integ/sagemaker/lineage/test_artifact.py]: > Do we need this manual retry if we're already wrapping this function in `retry()` ?

this test case is redundant, removed the test case

File: src/sagemaker/lineage/context.py (hunk 14 above)
Comment (yzhu0) [reply src/sagemaker/lineage/context.py]: removed

File: src/sagemaker/lineage/context.py (hunk 13 above)
Comment (yzhu0) [reply src/sagemaker/lineage/context.py]: > Wrong return type

fixed

File: src/sagemaker/lineage/artifact.py [hunk 18]
```diff
@@ -159,6 +159,51 @@ def downstream_trials(self, sagemaker_session=None) -> list:
...
+
+    def downstream_trials_v2(self) -> list:
+        """Use a lineage query to retrieve all downstream trials that use this artifact.
+
+        Returns:
+            [Trial]: A list of SageMaker `Trial` objects.
+        """
+        return self._trials(direction=LineageQueryDirectionEnum.DESCENDANTS)
+
+    def upstream_trials(self) -> List[str]:
```
Comment (staubhp) [inline src/sagemaker/lineage/artifact.py:172]: `List[str]` ?

File: src/sagemaker/lineage/artifact.py [hunk 19]
```diff
@@ -159,6 +159,51 @@ def downstream_trials(self, sagemaker_session=None) -> list:
...
+        """Use the lineage query to retrieve all upstream trials that use this artifact.
+
+        Returns:
+            [Trial]: A list of SageMaker `Trial` objects.
+        """
+        return self._trials(direction=LineageQueryDirectionEnum.ASCENDANTS)
+
+    def _trials(
+        self, direction: LineageQueryDirectionEnum = LineageQueryDirectionEnum.BOTH
+    ) -> List[str]:
```
Comment (staubhp) [inline src/sagemaker/lineage/artifact.py:182]: `List[str]`?

File: src/sagemaker/lineage/context.py [hunk 20]
```diff
@@ -382,11 +405,78 @@ def training_job_arns(
             training_job_arns.append(trial_component["Source"]["SourceArn"])
         return training_job_arns
 
+    def processing_jobs(
+        self, direction: LineageQueryDirectionEnum = LineageQueryDirectionEnum.ASCENDANTS
+    ):
```
Comment (staubhp) [inline src/sagemaker/lineage/context.py:410]: `def processing_jobs(
        self, direction: LineageQueryDirectionEnum = LineageQueryDirectionEnum.ASCENDANTS
    ) -> List[LineageTrialComponent]`

File: src/sagemaker/lineage/artifact.py [hunk 21]
```diff
@@ -522,3 +578,69 @@ def endpoint_contexts(
...
+
+
+class ImageArtifact(Artifact):
+    """A SageMaker lineage artifact representing an image.
+
+    Common model specific lineage traversals to discover how the image is connected
+    to other entities.
+    """
+
+    def datasets(self, direction: LineageQueryDirectionEnum):
```
[the remark this answers is not in the document]
Comment (yzhu0) [reply src/sagemaker/lineage/artifact.py:628]: > `def datasets(self, direction: LineageQueryDirectionEnum) -> List[Artifacts]:`

updated

File: src/sagemaker/lineage/artifact.py (hunk 18 above)
Comment (yzhu0) [reply src/sagemaker/lineage/artifact.py:172]: > `List[str]` ?

updated

File: src/sagemaker/lineage/artifact.py (hunk 19 above)
Comment (yzhu0) [reply src/sagemaker/lineage/artifact.py:182]: > `List[str]`?

updated

**Review by staubhp — APPROVED** [review]

**Comment by mufaddal-rohawala** [comment]
I see gaps in code coverage for unit tests, @yzhu0 can you help improve coverage?
https://codecov.io/gh/aws/sagemaker-python-sdk/commit/ddda8634a65e3f441414e9f9becf740185675bbe

**Review by mufaddal-rohawala — CHANGES_REQUESTED** [review]

**Comment by yzhu0** [comment]
> I see gaps in code coverage for unit tests, @yzhu0 can you help improve coverage? https://codecov.io/gh/aws/sagemaker-python-sdk/commit/ddda8634a65e3f441414e9f9becf740185675bbe

added unit tests for all new codes

**Review by mufaddal-rohawala — APPROVED** [review]

**Review by shreyapandit — APPROVED** [review]

**Pull request merged**